src/gui/screen_qr_error.cpp

```diff
@@ -4,36 +4,125 @@
-
-    info.SetBackColor(COLOR_RED_ALERT);
-    info.SetAlignment(Align_t::Center());
-    static const char hlp[] = "help.prusa3d.com";
-    info.SetText(string_view_utf8::MakeCPUFLASH((const uint8_t *)hlp));
+    : AddSuperWindow<screen_reset_error_data_t>()
+    , err_title(this, title_rect, is_multiline::no)
+    , err_description(this, descr_rect, is_multiline::yes)
+    , hand_icon(this, hand_rect, IDR_PNG_hand_qr)
+    , qr(this, QR_rect, 1) // error code is passed in the constructor
+    , help_txt(this, Rect16(0, 0, 0, 0), is_multiline::no)
+    , help_link(this, link_rect, is_multiline::no)
+    , qr_code_txt(this, qr_code_rect, is_multiline::no)
+    , fw_version_txt(this, fw_version_rect, is_multiline::no)
+    , signature_txt(this, signature_rect, is_multiline::no)
+    , appendix_txt(this, appendix_rect, is_multiline::no)
+    , sound_started(false) {
+
     SetRedLayout();
+    help_link.font = resource_font(IDR_FNT_SMALL);
+    qr_code_txt.font = resource_font(IDR_FNT_SMALL);
+    fw_version_txt.font = resource_font(IDR_FNT_SMALL);
+    signature_txt.font = resource_font(IDR_FNT_SMALL);
+    appendix_txt.font = resource_font(IDR_FNT_SMALL);
+
+    err_description.SetAlignment(Align_t::LeftTop());
+    hand_icon.SetAlignment(Align_t::LeftTop());
+    help_link.SetAlignment(Align_t::CenterTop());
+    qr_code_txt.SetAlignment(Align_t::CenterTop());
+    fw_version_txt.SetAlignment(Align_t::CenterTop());
+    signature_txt.SetAlignment(Align_t::CenterTop());
+    appendix_txt.SetAlignment(Align_t::CenterTop());
+
+    uint16_t error_code_short = dump_in_xflash_get_code();
+    uint16_t error_code = ERR_PRINTER_CODE * 1000 + error_code_short;
+    uint32_t i = 0;
+    uint32_t count = sizeof(error_list) / sizeof(err_t);
+
+    while (i < count && error_code_short != error_list[i].err_num) {
+        ++i;
+    }
+    if (i == count) {
+        /// no text found => leave blank screen
+        err_title.Hide();
+        err_description.Hide();
+        help_txt.Hide();
+        help_link.Hide();
+        qr.Hide();
+        qr_code_txt.Hide();
+    } else {
+        qr.SetQRHeader(error_code);
+        err_title.SetText(_(error_list[i].err_title));
+        err_description.SetText(_(error_list[i].err_text));
+
+        /// draw short URL
+        const char *qr_text = qr.GetQRShortText();
+        //help_txt.SetText(_(help_text));
+        help_txt.Hide();
+        help_link.SetText(_(qr_text));
+
+        if (eeprom_get_bool(EEVAR_DEVHASH_IN_QR)) {
+            static char p_code[PRINTER_CODE_SIZE + 1];
+            printerCode(p_code);
+            qr_code_txt.SetText(string_view_utf8::MakeRAM((const uint8_t *)p_code));
+        } else {
+            qr_code_txt.Hide();
+        }
+    }
+
+    /// draw footer information
+    /// fw version, hash, [fw signed], [appendix]
+    static char fw_version[14]; // intentionally limited to the number of practically printable characters without overwriting the nearby hash text
```

Comment:
don't use 14 directly
is not size of that string defined somewhere?
If not, you should define it and use it everywhere fw version is accessed. Even if you had to define new header with just a single constant

Reply:
`6d270b9`